From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Gyuyoung Kim <qkim@google.com>
Date: Wed, 28 Oct 2026 06:40:36 +0000
Subject: [M150-LTS] Avoid re-entry when draining pending error events

Fixed: 573264074
Change-Id: If8503bd2e9ca28fc5dd316d8aacb58ec168c6d38
Reviewed-on: https://chromium-review.googlesource.com/c/chromium/src/+/8622063
Reviewed-by: Michael Lippautz <mlippautz@chromium.org>
Reviewed-by: Nate Chapin <japhet@chromium.org>
Commit-Queue: Andrey Kosyakov <caseq@chromium.org>
Cr-Commit-Position: refs/heads/main@{#1724943}
(cherry picked from commit bd5ef225208e665908208a9123c5acdb0fea2993)

diff --git a/third_party/blink/renderer/core/execution_context/execution_context.cc b/third_party/blink/renderer/core/execution_context/execution_context.cc
index ba02ed9a0e2b3bae4f1be10cb178a889c10b3de0..2b0b6b0c43653075bd7d9c540d4c1af348211385 100644
--- a/third_party/blink/renderer/core/execution_context/execution_context.cc
+++ b/third_party/blink/renderer/core/execution_context/execution_context.cc
@@ -27,6 +27,7 @@
 
 #include "third_party/blink/renderer/core/execution_context/execution_context.h"
 
+#include "base/auto_reset.h"
 #include "base/metrics/histogram_functions.h"
 #include "base/tracing/protos/chrome_track_event.pbzero.h"
 #include "build/build_config.h"
@@ -283,20 +284,31 @@ void ExecutionContext::AddConsoleMessageImpl(
 void ExecutionContext::DispatchErrorEvent(
     ErrorEvent* error_event,
     SanitizeScriptErrors sanitize_script_errors) {
+  if (is_draining_pending_exceptions_) {
+    // Disregard errors reported while draining pending exceptions. This would
+    // help to avoid potential infinite recursion and, given this only
+    // happens with with inspector Runtime domain enabled, somewhat reduces
+    // devtools side effects.
+    return;
+  }
   if (in_dispatch_error_event_) {
     pending_exceptions_.push_back(error_event);
     return;
   }
 
   // First report the original exception and only then all the nested ones.
-  if (!DispatchErrorEventInternal(error_event, sanitize_script_errors))
+  if (!DispatchErrorEventInternal(error_event, sanitize_script_errors)) {
+    base::AutoReset auto_reset(&is_draining_pending_exceptions_, true);
     ExceptionThrown(error_event);
+  }
 
-  if (pending_exceptions_.empty())
-    return;
-  for (ErrorEvent* e : pending_exceptions_)
+  base::AutoReset auto_reset(&is_draining_pending_exceptions_, true);
+  auto pending_exceptions = std::move(pending_exceptions_);
+  for (ErrorEvent* e : pending_exceptions) {
     ExceptionThrown(e);
-  pending_exceptions_.clear();
+  }
+  // Assure handling pending exceptions doesn't add more pending exceptions.
+  CHECK(pending_exceptions_.empty());
 }
 
 bool ExecutionContext::DispatchErrorEventInternal(
diff --git a/third_party/blink/renderer/core/execution_context/execution_context.h b/third_party/blink/renderer/core/execution_context/execution_context.h
index 4a73a62e76bff486b61ea1c1b0f857fead038e91..1c78a66b85040404deeebb7c0f3c69f120600236 100644
--- a/third_party/blink/renderer/core/execution_context/execution_context.h
+++ b/third_party/blink/renderer/core/execution_context/execution_context.h
@@ -527,6 +527,7 @@ class CORE_EXPORT ExecutionContext : public Supplementable<ExecutionContext>,
   unsigned circular_sequential_id_;
 
   bool in_dispatch_error_event_;
+  bool is_draining_pending_exceptions_ = false;
   HeapVector<Member<ErrorEvent> > pending_exceptions_;
 
   mojom::FrameLifecycleState lifecycle_state_;
diff --git a/third_party/blink/web_tests/inspector-protocol/runtime/runtime-exception-thrown-nested-expected.txt b/third_party/blink/web_tests/inspector-protocol/runtime/runtime-exception-thrown-nested-expected.txt
new file mode 100644
index 0000000000000000000000000000000000000000..ea3debc866e94760852fae16dd5823744dc84daf
--- /dev/null
+++ b/third_party/blink/web_tests/inspector-protocol/runtime/runtime-exception-thrown-nested-expected.txt
@@ -0,0 +1,5 @@
+Tests that nested exceptions thrown while reporting an exception to Runtime domain do not crash.
+<no description>
+Error: error listener 1
+Error: error listener 2
+
diff --git a/third_party/blink/web_tests/inspector-protocol/runtime/runtime-exception-thrown-nested.js b/third_party/blink/web_tests/inspector-protocol/runtime/runtime-exception-thrown-nested.js
new file mode 100644
index 0000000000000000000000000000000000000000..1ed21a04ed2a3773f286fcf07d1cea073bdb64a2
--- /dev/null
+++ b/third_party/blink/web_tests/inspector-protocol/runtime/runtime-exception-thrown-nested.js
@@ -0,0 +1,45 @@
+(async function(/** @type {import('test_runner').TestRunner} */ testRunner) {
+  const {session, dp} = await testRunner.startHTML(
+      `
+    <script>
+      let nested = false;
+      Error.prepareStackTrace = (err, stack) => {
+        if (!nested) {
+          nested = true;
+          window.dispatchEvent(new Event('nested'));
+        }
+        return 'custom stack';
+      };
+
+      window.addEventListener('nested', () => {
+        throw new Error('from prepareStackTrace');
+      });
+
+      let firstError = true;
+      window.addEventListener('error', () => {
+        if (firstError) {
+          throw new Error('error listener 1');
+        }
+      });
+      window.addEventListener('error', () => {
+        if (firstError) {
+          firstError = false;
+          throw new Error('error listener 2');
+        }
+      });
+
+      window.addEventListener('trigger', () => {
+        throw 'initial';
+      });
+    </script>
+  `,
+      'Tests that nested exceptions thrown while reporting an exception to Runtime domain do not crash.');
+
+  await dp.Runtime.enable();
+  dp.Runtime.onExceptionThrown(event => {
+    testRunner.log(event.params.exceptionDetails.exception.description ||
+                   '<no description>');
+  });
+  await session.evaluate(`window.dispatchEvent(new Event('trigger'))`);
+  testRunner.completeTest();
+})
